be/src/storage/merge_iterator.cpp

```diff
@@ -317,27 +317,56 @@ inline Status MaskMergeIterator::do_get_next(Chunk* chunk, std::vector<RowSource
         DCHECK_GT(min_chunk.remaining_rows(), 0);
 
         size_t offset = min_chunk.compared_row();
-        // check whether |min_chunk| has overlapping with others.
         size_t min_chunk_num_rows = min_chunk._chunk->num_rows();
-        if (offset == 0 && _mask_buffer->has_same_source(child, min_chunk_num_rows)) {
-            if (rows == 0) {
-                chunk->swap_chunk(*min_chunk._chunk);
-                for (int i = 0; i < min_chunk_num_rows; ++i) {
-                    source_masks->emplace_back(_mask_buffer->current());
-                    _mask_buffer->advance();
+        size_t append_row_num = 0;
+        size_t max_same_source_count = _mask_buffer->max_same_source_count(child, min_chunk_num_rows);
+        if (max_same_source_count == min_chunk_num_rows) {
+            if (offset == 0) {
+                // all rows in |min_chunk| are from the same source chunk and |min_chunk|'s current offset is 0,
+                // so here we swap the whole min_chunk out.
+                if (rows == 0) {
+                    chunk->swap_chunk(*min_chunk._chunk);
+                    for (int i = 0; i < min_chunk_num_rows; ++i) {
+                        source_masks->emplace_back(_mask_buffer->current());
```

Comment:
why not insert batch

Reply:
RowSourceMask contains two fileds: `source_num` and `agg_flag`, here `source_num` are all equal to the min_chunk, however, `agg_flag` may be different.
So here i can't write source_masks->insert(source_masks.end(), min_chunk_num_rows, _mask_buffer->current());
Also, the underlying of _mask_buffer is a  FixedLengthColumn<uint16_t>, it only has `get(n)` API, i can only get one element each time.